docs/design-proposals/pv.md

```diff
@@ -0,0 +1,660 @@
+conducted.
+
+## Using block PVs in Virtlet
+
+As it turns out, the non-local block PVs aren't different from local
+block PVs from the CRI point of view. They're configured using
+`volumeDevices` section of the container spec in the pod and `volumes`
+section of the pod spec, and passed as `devices` section in the
+container config to `CreateContainer()` CRI call:
+
+```yaml
+  devices:
+  - container_path: /dev/testpvc
+    host_path: /var/lib/kubelet/pods/65b0c985-6e81-11e8-be27-769e6e14e66a/volumeDevices/kubernetes.io~local-volume/local-block-pv
+    permissions: mrw
+```
+
+Virtlet can use `host_path` to attach the device to the VM using a
+`DomainDisk`, and `container_path` to mount it inside the VM using
+cloud-init. The handling of local and non-local PVs doesn't differ
+on the CRI level.
+
+Supporting non-local PVs will automatically give Virtlet support for
+all the Kubernetes volume types that support the block mode, which
+include Ceph, FibreChannel, and the persistent disks on AWS, GCP and
+Azure, with the list probably growing larger in the future. It will
+also give automatic support for CSI plugins that support the block
+mode.  The caveat is that the block mode is Alpha as of Kubernetes
+1.10 and it wasn't checked for earlier Kubernetes versions.
+
+The use of block PVs will eliminate the need for custom flexvolumes at
+some point (after block volumes become GA and we stop supporting
+earlier Kubernetes versions). There's one caveat, with block PVs the
+Ceph RBDs will be mapped on the nodes by `kubelet`, instead of being
+consumed by qemu by the means of `librbd`. It's not clear though if
+this will be good or bad from the performance standpoint. If we'll
+still need custom volume types, flexvolumes may be replaced with
+[CSI](https://kubernetes.io/blog/2018/04/10/container-storage-interface-beta/).
+
+More advantages of using block PVs instead of custom flexvolumes
+include having VM pods differ even less from "plain" pods, and a
+possibility to make use automatic PV provisioning in future.
+
+There's also a possibility of using the block PVs (local or non-local)
+for the persistent rootfs. It's possible to copy the image onto PV
+upon the first use, and then have another pod reuse the PV after the
+original one is destroyed. For local PVs, the scheduler will always
+place the pod on the node where the local PV resides (this constitutes
+so called "gravity"). There's a problem with this approach, namely,
+there's no reliable way for a CRI implementation to find a PV that
+corresponds to a block device, so Virtlet will have to examine the
+contents of the PV to see if it's used for the first time. This also
+means that Virtlet will have hard time establishing the correspondence
+between PVs and the images that are applied to them (e.g. imagine a PV
+being used by a pod with different image later). It's possible to
+overcome these problems by either storing the metadata on the block
```

Comment:
You might want to look at CDI here and DataVolumes which came up in KubeVirt to solve the provisioning problem